src/Uno.UWP/Storage/Streams/Buffer.cs

```diff
@@ -1,27 +1,160 @@
 ﻿using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Windows.Foundation;
 
 namespace Windows.Storage.Streams
 {
 	public partial class Buffer : IBuffer
 	{
+		/// <summary>
+		/// A default length to use for buffer copy when none specified
+		/// </summary>
+		internal const int DefaultCapacity = 1024 * 1024; // 1M
+
+		private readonly Memory<byte> _data;
+		private uint _length;
+
+		internal static Buffer Cast(IBuffer impl)
+		{
+			if (impl is Buffer buffer)
+			{
+				return buffer;
+			}
+			else if (impl is null)
+			{
+				throw new ArgumentNullException(nameof(impl));
+			}
+			else
+			{
+				throw new NotSupportedException("This type of buffer is not supported.");
+			}
+		}
+
 		public Buffer(uint capacity)
 		{
-			Data = new byte[capacity];
+			_data = new Memory<byte>(new byte[capacity]);
 		}
 
 		internal Buffer(byte[] data)
 		{
-			Data = data;
+			_data = new Memory<byte>(data);
+			Length = Capacity;
 		}
 
-		internal byte[] Data { get; }
+		internal Buffer(Memory<byte> data)
+		{
+			_data = data;
+			Length = (uint)data.Length;
+		}
 
-		public uint Capacity => (uint)Data.Length;
+		public uint Capacity => (uint)_data.Length;
 
 		public uint Length
 		{
-			get => (uint)Data.Length;
-			set => throw new NotSupportedException();
+			get => _length;
+			set
+			{
+				if (value > Capacity) throw new ArgumentOutOfRangeException(nameof(value), "Length cannot be greater than capacity.");
```

Comment:
Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6307767)